cluster-autoscaler/cloudprovider/mcm/mcm_cloud_provider.go

```
@@ -281,152 +254,211 @@ func ReferenceFromProviderID(m *McmManager, id string) (*Ref, error) {
-func (machinedeployment *MachineDeployment) DecreaseTargetSize(delta int) error {
-	klog.V(0).Infof("Received request to decrease target size of machine deployment %s by %d", machinedeployment.Name, delta)
+func (machineDeployment *MachineDeployment) DecreaseTargetSize(delta int) error {
+	klog.V(0).Infof("Received request to decrease target size of machine deployment %s by %d", machineDeployment.Name, delta)
 	if delta >= 0 {
 		return fmt.Errorf("size decrease size must be negative")
 	}
-	size, err := machinedeployment.mcmManager.GetMachineDeploymentSize(machinedeployment)
+	machineDeployment.scalingMutex.Lock()
+	defer machineDeployment.scalingMutex.Unlock()
+	size, err := machineDeployment.mcmManager.GetMachineDeploymentSize(machineDeployment)
 	if err != nil {
 		return err
 	}
 	decreaseAmount := int(size) + delta
-	if decreaseAmount < machinedeployment.minSize {
-		klog.Warningf("Cannot go below min size= %d for machineDeployment %s, requested target size= %d . Setting target size to min size", machinedeployment.minSize, machinedeployment.Name, size+int64(delta))
-		decreaseAmount = machinedeployment.minSize
+	if decreaseAmount < machineDeployment.minSize {
+		klog.Warningf("Cannot go below min size= %d for machineDeployment %s, requested target size= %d . Setting target size to min size", machineDeployment.minSize, machineDeployment.Name, size+int64(delta))
+		decreaseAmount = machineDeployment.minSize
+	}
+	return machineDeployment.mcmManager.retry(func(ctx context.Context) (bool, error) {
+		return machineDeployment.mcmManager.SetMachineDeploymentSize(ctx, machineDeployment, int64(decreaseAmount))
+	}, "MachineDeployment", "update", machineDeployment.Name)
+}
+
+// Refresh resets the priority annotation for the machines that are not present in machines-marked-by-ca-for-deletion annotation on the machineDeployment
+func (machineDeployment *MachineDeployment) Refresh() error {
+	machineDeployment.scalingMutex.Lock()
```

why should a scaling mutex be used in `Refresh`?
Scaling mutex by its very name signifies mutex meant to be taken for any scaling event and not when its trying to read/get machine deployments.

Also can this result in the next reconcile of `RunOnce` to be stuck because it cannot go beyond `CloudProvider.Refresh` call?

---

We need a mutex when running scaling operations or modifying attributes that directly affect the scaling operation like the newly introduced annotation. There is not any doubt about this as we HAVE to prevent concurrent scaleups and scsledowns due to the fact that our MCM scaling is async and that multiple Go routines can perform scale downs, leading to non-deterministic behaviour.

You have a point that `RunOnce` should return an error and not block if the mutex is already acquired. Will change the code to check if mutex is acquired and if so return an informative error. Will also test this change.

---

Modified to take a `TryLock` in `NodeGroupImpl.Refresh()`. This is also a limited fix since it will block CA from changing other NodeGroups but won't block go-routines. I don't see any better way with current CA architecture. One possible improvement is to introduce a mutex with a timeout.  If you find a better way, please let me know.